### PR TITLE
[tool_requires][best-practices] Using protobuf

### PR DESCRIPTION
**Pull request by franramirez688**
Depends on https://github.com/conan-io/examples2/pull/97

* `myaddresser`: using protobuf as `requires` and `tool_requires` at the same time (cross-building as well)

Related docs: https://github.com/conan-io/docs/pull/3222

### CODE REVIEW

**Review by memsharded — COMMENTED** [review]
Use "mylibrary`` or better something with meaning, like "myserver", as the example, with its ``test_package``, instead of using an extra ``consumer`` application package.

**Review by memsharded — APPROVED** [review]
It is looking good, but it must wait until the other PR for ``tool-requires`` is merged, then updated. Or just clean the other PR changes.

File: examples/graph/tool_requires/using_protobuf/myserver/test_package/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.15)
+project(PackageTest CXX)
+
+find_package(protobuf CONFIG REQUIRED)
```
Comment (memsharded) [inline examples/graph/tool_requires/using_protobuf/myserver/test_package/CMakeLists.txt:4]: protobuf not necessary as a direct dependency? Typically only ``myserver`` needs a ``find_package()``

File: examples/graph/tool_requires/using_protobuf/myserver/src/myserver.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
+#include <iostream>
+#include "myserver.h"
+
+
+
+void myserver(){
```
Comment (memsharded) [inline examples/graph/tool_requires/using_protobuf/myserver/src/myserver.cpp:6]: no actual usage of protobuf?

File: examples/graph/tool_requires/using_protobuf/myserver/src/myserver.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    #ifdef _M_X64
+    std::cout << "  myserver/1.0: _M_X64 defined\n";
+    #endif
+
+    #ifdef _M_IX86
+    std::cout << "  myserver/1.0: _M_IX86 defined\n";
+    #endif
+
+    #ifdef _M_ARM64
+    std::cout << "  myserver/1.0: _M_ARM64 defined\n";
```
Comment (memsharded) [inline examples/graph/tool_requires/using_protobuf/myserver/src/myserver.cpp:25]: All this extra output can be removed for simplicity

File: examples/graph/tool_requires/using_protobuf/myserver/test_package/src/example.cpp [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#include <iostream>
+#include <fstream>
+#include <string>
+#include "addressbook.pb.h"
+
+
+int main(int argc, char* argv[]) {
```
Comment (memsharded) [inline examples/graph/tool_requires/using_protobuf/myserver/test_package/src/example.cpp:7]: ``test_package`` of ``myserver`` doesnt' test ``myserver``?

**Review by memsharded — APPROVED** [review]

File: examples/graph/tool_requires/different_versions/run_example.py [hunk 5]
```diff
@@ -8,4 +8,4 @@
 
     output = run("conan create wine")
     assert "MYGCC=1.0!!" in output
-    assert "MYGCC=2.0!!" in output
+    assert "MYGCC=2.0!!" in output
```
Comment (franramirez688) [inline examples/graph/tool_requires/different_versions/run_example.py:11]:
```suggestion
    assert "MYGCC=2.0!!" in output

```

File: examples/graph/tool_requires/using_protobuf/myaddresser/apple-arch-armv8 [hunk 6]
```diff
@@ -0,0 +1,9 @@
+[settings]
+arch=armv8
+build_type=Release
+compiler=apple-clang
+compiler.cppstd=gnu17
+compiler.libcxx=libc++
+compiler.version=13.0
```
Comment (czoido) [inline examples/graph/tool_requires/using_protobuf/myaddresser/apple-arch-armv8:7]:
```suggestion
compiler.version=13
```

File: examples/graph/tool_requires/using_protobuf/myaddresser/apple-arch-x86_64 [hunk 7]
```diff
@@ -0,0 +1,8 @@
+[settings]
+arch=x86_64
+build_type=Release
+compiler=apple-clang
+compiler.cppstd=gnu17
+compiler.libcxx=libc++
+compiler.version=13.0
```
Comment (czoido) [inline examples/graph/tool_requires/using_protobuf/myaddresser/apple-arch-x86_64:7]:
```suggestion
compiler.version=13
```

**Review by czoido — APPROVED** [review]

File: examples/graph/tool_requires/using_protobuf/run_example.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+assert "myaddresser(): created a person with id 1337" in output
+# Cross-building to Intel arch
+output = run("conan create myaddresser --build missing -pr:b myaddresser/apple-arch-armv8 -pr:h myaddresser/apple-arch-x86_64")
+assert "myaddresser(): created a person with id 1337" not in output
+
+# Assert the architectures for both testing examples created
+output = run("file myaddresser/test_package/build/apple-clang-13.0-armv8-gnu17-release/example")
+assert "64-bit executable arm64" in output
+output = run("file myaddresser/test_package/build/apple-clang-13.0-x86_64-gnu17-release/example")
+assert "64-bit executable x86_64" in output
```
Comment (czoido) [inline examples/graph/tool_requires/using_protobuf/run_example.py:21]:
```suggestion
output = run("file myaddresser/test_package/build/apple-clang-13-armv8-gnu17-release/example")
assert "64-bit executable arm64" in output
output = run("file myaddresser/test_package/build/apple-clang-13-x86_64-gnu17-release/example")
assert "64-bit executable x86_64" in output
```

**Pull request merged**